### PR TITLE
Add GetChassisThermals command to Chassis category of redfish_facts

### LINKED ISSUE #54231: redfish_facts- Chassis - GetChassisThermals
##### SUMMARY

This feature would implement a GetChassisThermals command for the Chassis category of redfish_facts, and would retrieve temperature related properties from the Chassis/Thermal field for each sensor available.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

redfish_facts
##### ADDITIONAL INFORMATION

```yaml

```

### PR DESCRIPTION
**Pull request by xmadsen**
##### SUMMARY

This pull request adds a GetChassisThermals command to the Chassis category of redfish_facts. It retrieves from the Thermal/Temperatures field various temperature-related data for each sensor populated therein.

Fixes #54231 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

redfish_facts

##### ADDITIONAL INFORMATION

```yaml
- name: Test Redfish modules
  hosts: localhost
  gather_facts: false
  vars:
    baseuri: 10.243.5.31
    user: USERID
    password: PASSW0RD

  tasks:
    - name: Get Chassis Thermals
      redfish_facts:
        category: Chassis
        command: GetChassisThermals
        baseuri: "{{ baseuri }}"
        username: "{{ user }}"
        password: "{{ password }}"
```

```paste below
ansible-playbook 2.8.0.dev0
  config file = None
  configured module search path = [u'/home/xander/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /mnt/c/Users/amadsen/Coding/lenovo-redfish-automated-testing/ansible/lib/ansible
  executable location = /mnt/c/Users/amadsen/Coding/lenovo-redfish-automated-testing/ansible/bin/ansible-playbook
  python version = 2.7.12 (default, Nov 12 2018, 14:36:49) [GCC 5.4.0 20160609]
No config file found; using defaults
host_list declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not
match 'all'


PLAYBOOK: test_redfish_facts.1.yml **********************************************************************************1 plays in ../test_redfish_facts.1.yml

PLAY [Test Redfish modules] *****************************************************************************************META: ran handlers

TASK [Get Chassis Thermals] *****************************************************************************************task path: /mnt/c/Users/amadsen/Coding/lenovo-redfish-automated-testing/test_redfish_facts.1.yml:10
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: xander
<127.0.0.1> EXEC /bin/sh -c 'echo ~xander && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880 `" && echo ansible-tmp-1553609768.02-132527517191880="` echo /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880 `" ) && sleep 0'
Using module file /mnt/c/Users/amadsen/Coding/lenovo-redfish-automated-testing/ansible/lib/ansible/modules/remote_management/redfish/redfish_facts.py
<127.0.0.1> PUT /home/xander/.ansible/tmp/ansible-local-8740sZYjwN/tmpESZy_z TO /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880/AnsiballZ_redfish_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880/ /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880/AnsiballZ_redfish_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880/AnsiballZ_redfish_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/xander/.ansible/tmp/ansible-tmp-1553609768.02-132527517191880/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => {
    "ansible_facts": {
        "redfish_facts": {
            "thermals": {
                "entries": [
                    {
                        "Name": "Ambient Temp", 
                        "PhysicalContext": "Intake", 
                        "ReadingCelsius": 24, 
                        "UpperThresholdCritical": 47, 
                        "UpperThresholdFatal": 50
                    }, 
                    {
                        "Name": "Exhaust Temp", 
                        "PhysicalContext": "Exhaust", 
                        "ReadingCelsius": 27
                    }, 
                    {
                        "Name": "CPU1 Temp", 
                        "PhysicalContext": "CPU", 
                        "ReadingCelsius": 26
                    }, 
                    {
                        "Name": "CPU2 Temp", 
                        "PhysicalContext": "CPU", 
                        "ReadingCelsius": 24
                    }, 
                    {
                        "Name": "CPU1 DTS", 
                        "PhysicalContext": "CPU", 
                        "ReadingCelsius": -510, 
                        "UpperThresholdCritical": -2, 
                        "UpperThresholdFatal": 0
                    }, 
                    {
                        "Name": "CPU2 DTS", 
                        "PhysicalContext": "CPU", 
                        "ReadingCelsius": -510, 
                        "UpperThresholdCritical": -2, 
                        "UpperThresholdFatal": 0
                    }, 
                    {
                        "Name": "DIMM 5 Temp", 
                        "PhysicalContext": "Memory", 
                        "ReadingCelsius": 27
                    }, 
                    {
                        "Name": "DIMM 17 Temp", 
                        "PhysicalContext": "Memory", 
                        "ReadingCelsius": 27
                    }, 
                    {
                        "Name": "PCH Temp", 
                        "PhysicalContext": "ComputeBay", 
                        "ReadingCelsius": 49
                    }
                ], 
                "ret": true
            }
        }
    }, 
    "changed": false, 
    "invocation": {
        "module_args": {
            "baseuri": "10.243.5.31", 
            "category": [
                "Chassis"
            ], 
            "command": [
                "GetChassisThermals"
            ], 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "username": "USERID"
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************************************************localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

### CODE REVIEW

**Comment by xmadsen** [comment]
@samerhaj

**Comment by xmadsen** [comment]
it looks like shippable can't clone the ansible repo for some reason?

**Comment by samerhaj** [comment]
Some feedback:

* The following properties should be included if they are present and not null, since they have useful information:
LowerThresholdFatal
LowerThresholdNonCritical
LowerThresholdCritical
UpperThresholdFatal
UpperThresholdNonCritical
UpperThresholdCritical
MinReadingRangeTemp
MaxReadingRangeTemp
SensorNumber

in addition to the ones you added: Name, PhysicalContext, and ReadingCelsius.

* Since you are combining Temperatures from multiple chassis, and also since a single Chassis will have Temp sensors spanning multiple components, the RelatedItems[] is a key piece of data that need to be included in the output to map the reading to the correct component.

**Comment by samerhaj** [comment]
Looks good! Just fix the trailing whitespaces, and shipit!

**Comment by samerhaj** [comment]
shipit

**Review by billdodd — APPROVED** [review]
Reviewed and tested. Looks good to me.

shipit

**Comment by jose-delarosa** [comment]
Looks good @xmadsen, might need to rebase your branch. Thank you.

shipit

**Review by billdodd — APPROVED** [review]
Approving again after rebase

shipit

**Comment by jose-delarosa** [comment]
Looks good.

shipit

**Comment by gundalow** [comment]
Another rebase needed :(

**Comment by gundalow** [comment]
If someone could poke me on IRC once this has been rebased I'll get it merged. I get too much GitHub email to keep on top of.

**Comment by billdodd** [comment]
@gundalow - Thanks. I'll ping you on IRC once this has been rebased.

@xmadsen - For the 3(?) open chassis related PRs that need rebasing and/or updates, it might be best to do one of them, then wait for it to be merged. Then do the next one, etc. Otherwise, they'll probably keep conflicting with each other in redfish_facts.py.

**Review by billdodd — APPROVED** [review]
Re-reviewed after rebase. Still looks good.

shipit

**Comment by jose-delarosa** [comment]
shipit

**Comment by xmadsen** [comment]
@gundalow this is ready to be merged

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged into `devel` for release in Ansible 2.9
It's great to see such a community of reviewers around the Remote Management PRs

**Pull request merged**
